ducktape/tests/runner_client.py

```diff
@@ -102,82 +107,105 @@ def run(self):
-            self.test_context = None
-            self.test = None
+        all_services = ServiceRegistry()
+
+        num_runs = 1
+        while test_status == FAIL and num_runs <= self.deflake_num:
+            self.log(logging.INFO, "on run {}/{}".format(num_runs, self.deflake_num))
+            try:
+                # Results from this test, as well as logs will be dumped here
+                mkdir_p(TestContext.results_dir(self.test_context, self.test_index))
+                # Instantiate test
+                self.test = self.test_context.cls(self.test_context)
+
+                self.log(logging.DEBUG, "Checking if there are enough nodes...")
+                min_cluster_spec = self.test.min_cluster_spec()
+                os_to_num_nodes = {}
+                for node_spec in min_cluster_spec:
+                    if not os_to_num_nodes.get(node_spec.operating_system):
+                        os_to_num_nodes[node_spec.operating_system] = 1
+                    else:
+                        os_to_num_nodes[node_spec.operating_system] = os_to_num_nodes[node_spec.operating_system] + 1
+                for (operating_system, node_count) in iteritems(os_to_num_nodes):
+                    num_avail = len(list(self.cluster.all().nodes.elements(operating_system=operating_system)))
+                    if node_count > num_avail:
+                        raise RuntimeError(
+                            "There are not enough nodes available in the cluster to run this test. "
+                            "Cluster size for %s: %d, Need at least: %d. Services currently registered: %s" %
+                            (operating_system, num_avail, node_count, self.test_context.services))
+
+                # Run the test unit
+                start_time = time.time()
+                self.setup_test()
+
+                data = self.run_test()
+
+                test_status = PASS if num_runs == 1 else FLAKY
+                self.log(logging.INFO, "{} TEST".format(test_status.to_json()))
+
+            except BaseException as e:
+                # mark the test as failed before doing anything else
+                test_status = FAIL
+                err_trace = self._exc_msg(e)
+                summary.append(err_trace)
+                if num_runs != self.deflake_num:
+                    summary.append("~" * max(len(l) for l in err_trace.split('\n')) + "\n")
+                self.log(logging.INFO, "FAIL: " + err_trace)
+
+            finally:
+                for service in self.test_context.services:
+                    if self.deflake_num > 1:
+                        service.service_id_factory = MultiRunServiceIdFactory(service, num_runs)
+                    all_services.append(service)
+
+                self.teardown_test(teardown_services=not self.session_context.no_teardown, test_status=test_status)
+
+                stop_time = time.time()
+
+                if hasattr(self.test_context, "services"):
```

Comment:
move it before the teardown, otherwise service registry is empty by the time you get here (its wiped in teardown I believe)

Reply:
Ive moved the free_nodes out of teardown, and placed it after here, so that way we can capture errors from cleanup as well